drivers/spi/spi_context.h

```diff
@@ -173,29 +184,47 @@ static inline int spi_context_wait_for_completion(struct spi_context *ctx)
 
 	if (wait) {
 		k_timeout_t timeout;
+		uint32_t timeout_ms;
 
 		/* Do not use any timeout in the slave mode, as in this case
 		 * it is not known when the transfer will actually start and
 		 * what the frequency will be.
 		 */
 		if (IS_ENABLED(CONFIG_SPI_SLAVE) && spi_context_is_slave(ctx)) {
 			timeout = K_FOREVER;
+			timeout_ms = UINT32_MAX;
 		} else {
 			uint32_t tx_len = spi_context_total_tx_len(ctx);
 			uint32_t rx_len = spi_context_total_rx_len(ctx);
-			uint32_t timeout_ms;
 
 			timeout_ms = MAX(tx_len, rx_len) * 8 * 1000 /
 				     ctx->config->frequency;
 			timeout_ms += CONFIG_SPI_COMPLETION_TIMEOUT_TOLERANCE;
 
 			timeout = K_MSEC(timeout_ms);
 		}
-
+#ifdef CONFIG_MULTITHREADING
 		if (k_sem_take(&ctx->sync, timeout)) {
 			LOG_ERR("Timeout waiting for transfer complete");
 			return -ETIMEDOUT;
 		}
+#else
+		const uint32_t tms = k_uptime_get_32();
+		unsigned int key = irq_lock();
+
+		while (!ctx->ready && (k_uptime_get_32() - tms < timeout_ms)) {
```

Comment:
I really don't love this when we have a much nicer answer to using buses without threads right next store, and nrf even has a driver supporting it for i2c!